### ISSUE TITLE
Allow connecting with Unix Socket to Rest Server

### ISSUE THREAD

**Issue by shyim** [restic/restic#4287]
Output of `restic version`
--------------------------

```
restic 0.15.1 compiled with go1.20.2 on linux/arm64
```

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

Allow connecting with Unix Socket to Rest Server instead of normal IP

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I would like to use the rest server with sockets to have an cleaner setup with systemd services and don't have to think about ports

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

The best backup tool I had. It just works and is simple to use. Thanks guys!

**Comment by MichaelEischer** [comment]
It should be possible to allow the usage of `rest:unix://some/path`. But someone would have to figure out how to configure the http roundtripper accordingly.

**Comment by Enrico204** [comment]
I think that something like this is sufficient:

```go
client := http.Client{
  Transport: &http.Transport{
    DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
      return net.DialContext(ctx, "unix", "/socket/path") 
    }
  }
}
```

I suspect that something similar is available on the server side for `rest-server`.

**Comment by MichaelEischer** [comment]
Thanks, that looks like it could work. Anyone up for a PR?

**Comment by ae-govau** [comment]
This may be of interest to us. I did some work today on a PR to `rest-server` to add support for listening on a unix socket.

The change itself was small, but when adding tests, I had to refactor `main.go` a little to get away from using global variables, which makes the change look much bigger than it really is. I'll post shortly for feedback, and link to this issue.

(and then I'll make a corresponding change here - that snippet from @Enrico204 makes it easy)
